pipeline/processor/postprocessor.py

```diff
@@ -0,0 +1,634 @@
+from collections import defaultdict
+from copy import deepcopy
+from typing import List, Dict, Optional
+
+import pandas as pd
+from english_words import english_words_set
+
+from pipeline.extract.document import Document, TextBlock
+
+
+class PostProcessor:
+    @staticmethod
+    def _minimal_bounding_box(coords: List[list]) -> list:
+        """
+        Return the minimally enclosing bounding box of bounding boxes.
+
+        Args: coords: A list of coordinates for each bounding box formatted [x1,y1,x2,y2] with the bottom left as the
+        origin.
+
+        Returns:
+            A list of coordinates for the minimally enclosing bounding box for all input bounding boxes.
+
+        """
+        x_min = min(coord[0][0] for coord in coords)
+        y_min = min(coord[1][1] for coord in coords)
+        x_max = max(coord[1][0] for coord in coords)
+        y_max = max(coord[3][1] for coord in coords)
+        return [[x_min, y_min], [x_max, y_min], [x_min, y_max], [x_max, y_max]]
+
+
+class HyphenationPostProcessor:
+    """
+    Post processor to join words that are separated into separate blocks due
+    to hyphenation at line breaks.
+    """
+
+    @staticmethod
+    def _rewrap_hyphenated_words(li: list) -> list:
+        """
+        Reorganise a list of strings so that word fragments separated
+        by hyphens or em dashes to start new lines are joined into a single
+        word if the word fragments have no meaning by themselves.
+
+        Args:
+            li: List of strings/sentences.
+
+        Returns:
+            List of strings/sentences with hyphenated words joined into a single word.
+
+        """
+        current = None
+        for ix, l in enumerate(li):
+            regex_match = re.search(r"\w+(-|–){1}$", l)
+
+            if current:
+                word_fragment = current.rstrip("-")
+                # TODO: Handle non-English words
+                if word_fragment in english_words_set:
+                    # Check if the word with hyphen removed is an english
+                    # word and if it is, make this the first word of the newline
+                    # without the hyphen (e.g. repair-ing)
+                    newline_first_word = word_fragment + l.split(" ")[0].lstrip("-")
+                    if newline_first_word in english_words_set:
+                        li[ix] = newline_first_word + l.split(" ")[1:]
+                    # Otherwise, keep the hyphenation but put it on a newline e.g. post-processing.
+                    else:
+                        li[ix] = word_fragment + "-" + l
+                else:
+                    li[ix] = word_fragment + l
+                # Reset.
+                current = None
+            if regex_match:
+                # Strip matching regex from the end of the string.
+                li[ix] = l[: regex_match.start()]
+                current = regex_match[0]
+        return li
+
+    def process(self, contents: dict) -> dict:
```

Comment:
For the `process` methods, I'm thinking that even if the underlying logic operates on dicts, they should still take and receive `Document` objects? This way, if we add field validation to the `Document` class it's automatically built into the postprocessors.

It should be pretty easy to define a `Document.to_dict()` method based on `Document.save_json()`. I'm not sure about the effort involved in a `Document.from_dict()` classmethod though, or how much effort it would take to input and output `Document` objects through your code. Happy for you to make the call on this.

Reply:
Easy fix, added a bunch of class methods to handle.